### PR TITLE
Adding a devcontainer with Redis and a Ollama server 

### PR DESCRIPTION
**Pull request by ProKil**
This PR introduces an experimental dev container which would allow developers to easily set up a clean environment with a Redis server and an LLM server running in the local environment with minimal memory footprint (<4GB) so that it could run easily on laptop and codespaces. 

## 📑 Description

1. .devcontainer/devcontainer.json a specification of a container with Python, Redis, and Ollama support. It would automatically install poetry and sotopia, and serve Llama 3.2 1b on start. 
2. An example for using custom models, which can also serve a sanity check for the devcontainer. 
3. Fix agenerate API to use the `model_name` for reformatting. 
4. Updated the doc for contributors. 

Out of scope: 

1. Use the devcontainer in CI. This will be more of a breaking change which could be a separate PR.
2. Pre-load sotopia data into redis. The technical implementation needs more discussion.

## ✅ Checks

- [x] My pull request adheres to the code style of this project
- [x] My code requires changes to the documentation
- [x] I have updated the documentation as required
- [x] All the tests have passed
- [x] Branch name follows `type/descript` (e.g. `feature/add-llm-agents`)
- [x] Ready for code review

### CODE REVIEW

File: examples/generation_api/minimal_demo.py
Comment (XuhuiZhou) [inline examples/generation_api/minimal_demo.py:1]: why remove this?

File: examples/generation_api/custom_model.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# Example: Generate (2~7) random numbers using Llama3.2 model served by ollama
+# To run this example, you can either
+# 1. Use the sotopia devcontainer and run the following command in the terminal:
+#        poetry run python examples/generation_api/custom_model.py
+#    This example can also serve a sanity check for your devcontainer setup.
+# OR 2. after installing sotopia, install ollama, and then:
```
Comment (XuhuiZhou) [inline examples/generation_api/custom_model.py:6]: I don't think it's clear to me how we connect with the customized model here. Can we add things to the doc?

File: sotopia/generation_utils/generate.py [hunk 2]
```diff
@@ -777,7 +765,7 @@ def process_history(
 async def agenerate_init_profile(
     model_name: str,
     basic_info: dict[str, str],
-    bad_output_process_model: str = DEFAULT_BAD_OUTPUT_PROCESS_MODEL,
```
Comment (XuhuiZhou) [inline sotopia/generation_utils/generate.py:780]: why remove the default?

**Review by XuhuiZhou — CHANGES_REQUESTED** [review]

File: docs/pages/contribution/contribution.md [hunk 3]
```diff
@@ -60,7 +60,17 @@ git push origin main
 
 ### 4. Set up the Development Environment
 
-?? @ProKil
+We recommend using DevContainers to set up your development environment.
+
+#### Using VSCode
```
Comment (XuhuiZhou) [inline docs/pages/contribution/contribution.md:65]: This whole setup is wayyyy too VSCode centric, I don't know if people who don't use vscode would appreciate it

**Review by XuhuiZhou — CHANGES_REQUESTED** [review]

File: docs/pages/contribution/contribution.md (hunk 3 above)
Comment (ProKil) [reply docs/pages/contribution/contribution.md:65]: I don't have other softwares to test. PyCharm is also supported by Dev Containers. One can contribute their usage to the doc.

File: sotopia/generation_utils/generate.py (hunk 2 above)
Comment (ProKil) [reply sotopia/generation_utils/generate.py:780]: To use custom model without OpenAI keys, the original way doesn't work. Now we use the the model as the default reformat model, but allow the user to use other models for reformatting.

File: examples/generation_api/custom_model.py (hunk 1 above)
Comment (ProKil) [reply examples/generation_api/custom_model.py:6]: It is set up directly in devcontainer

File: sotopia/generation_utils/generate.py (hunk 2 above)
Comment (XuhuiZhou) [reply sotopia/generation_utils/generate.py:780]: would this break the existing running script that does not specify reformatting models? can we still have a default one?

**Review by XuhuiZhou — APPROVED** [review]

File: sotopia/generation_utils/generate.py (hunk 2 above)
Comment (ProKil) [reply sotopia/generation_utils/generate.py:780]: yeah, the default one would be the one used for generating the output for the first time.

**Pull request merged**